File: 0043-multiply-strings/0043-multiply-strings.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") return "0";
        
        int n = num1.size(), m = num2.size();
        // The result of multiplying two numbers of size n and m 
        // will have at most n + m digits.
        vector<int> pos(n + m, 0);
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                int mul = (num1[i] - '0') * (num2[j] - '0');
                int p1 = i + j, p2 = i + j + 1;
                int sum = mul + pos[p2];
                
                pos[p1] += sum / 10;
                pos[p2] = sum % 10;
            }
        }
        
        string res = "";
        for (int p : pos) {
            if (!(res.empty() && p == 0)) {
                res.push_back(p + '0');
            }
        }
        
        return res.empty() ? "0" : res;
    }
};

File: 0043-multiply-strings/0043-multiply-strings.py
class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        if num1 == "0" or num2 == "0":
            return "0"
        
        # The result of multiplying two numbers of length m and n
        # has at most m + n digits.
        m, n = len(num1), len(num2)
        res = [0] * (m + n)
        
        # Traverse from right to left
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                mul = int(num1[i]) * int(num2[j])
                
                # Position in result array
                p1, p2 = i + j, i + j + 1
                total = mul + res[p2]
                
                # Add to result array and handle carry
                res[p2] = total % 10
                res[p1] += total // 10
                
        # Convert result array back to string
        # Skip leading zero if exists
        result_str = "".join(map(str, res))
        return result_str.lstrip("0")